File: isValid.cpp
//use hashmap and stack
//Time complexity -> O(n)
//Space complexity -> O(n)
#include <unordered_map>
class Solution {
public:
    bool isValid(string s) {
        std::unordered_map<char,char> hashmap = 
        {
            { '}', '{' },
            { ')', '(' },
            { ']', '[' }
        };
        std::stack<int> stack;

        for(char c : s) {
            if (hashmap.count(c)) {
                if(!stack.empty() && stack.top() == hashmap[c]){
                    stack.pop();
                }
                else{
                    return false;
                }
            }
            else{
            //this means that it's an open character
            stack.push(c);                
            }
        }
        return stack.empty();
    }
};

File: isValid.py
#use stack and hashmap
#time complexity -> O(n) because stack + hashmap was used
#space complexity -> O(n) used a for loop

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        hashmap = { '}': '{',   ')': '(',   ']': '[' }

        for char in s:
            if char in hashmap:
                #now check if the stack has any char and if the top of the stack is the opening
                #char of the current char
                if stack and stack[-1] == hashmap[char]:
                    stack.pop()
                else:
                    return False
            #else represents open characters
            else:
                stack.append(char)

        if not stack:
            return True
        else:
            return False
          

#########################################################
#########################################################